File: packages/pc-keiba-viewer-plugin/src/mcp-stdio-proxy.ts
// Run with bun.

export interface McpProxyConfig {
  accessClientId: string;
  accessClientSecret: string;
  mcpAuthToken: string;
  mcpUrl: string;
}

export interface McpProxyEnv {
  MCP_AUTH_TOKEN?: string;
  PC_KEIBA_ACCESS_CLIENT_ID?: string;
  PC_KEIBA_ACCESS_CLIENT_SECRET?: string;
  PC_KEIBA_VIEWER_MCP_URL?: string;
}

const HTTPS_PROTOCOL: string = "https:";
const USER_AGENT: string = "pc-keiba-viewer-plugin/1.0";
const CONTENT_LENGTH_HEADER: string = "Content-Length:";

export const loadMcpProxyConfig = (env: McpProxyEnv): McpProxyConfig | string => {
  const mcpUrl = env.PC_KEIBA_VIEWER_MCP_URL?.trim() ?? "";
  const mcpAuthToken = env.MCP_AUTH_TOKEN?.trim() ?? "";
  const accessClientId = env.PC_KEIBA_ACCESS_CLIENT_ID?.trim() ?? "";
  const accessClientSecret = env.PC_KEIBA_ACCESS_CLIENT_SECRET?.trim() ?? "";
  if (mcpUrl.length === 0) {
    return "PC_KEIBA_VIEWER_MCP_URL is required (absolute https URL ending with /mcp)";
  }
  if (mcpAuthToken.length === 0) {
    return "MCP_AUTH_TOKEN is required";
  }
  if (accessClientId.length === 0) {
    return "PC_KEIBA_ACCESS_CLIENT_ID is required";
  }
  if (accessClientSecret.length === 0) {
    return "PC_KEIBA_ACCESS_CLIENT_SECRET is required";
  }
  try {
    const url = new URL(mcpUrl);
    if (url.protocol !== HTTPS_PROTOCOL) {
      return "PC_KEIBA_VIEWER_MCP_URL must be https";
    }
    if (url.username.length > 0 || url.password.length > 0) {
      return "PC_KEIBA_VIEWER_MCP_URL must not include userinfo";
    }
    if (url.hash.length > 0) {
      return "PC_KEIBA_VIEWER_MCP_URL must not include a fragment";
    }
    if (!url.pathname.endsWith("/mcp")) {
      return "PC_KEIBA_VIEWER_MCP_URL path must end with /mcp";
    }
  } catch {
    return "PC_KEIBA_VIEWER_MCP_URL is not a valid URL";
  }
  return {
    accessClientId,
    accessClientSecret,
    mcpAuthToken,
    mcpUrl,
  };
};

export const extractStdioMessages = (buffer: string): { messages: string[]; rest: string } => {
  const messages: string[] = [];
  let rest = buffer;
  while (rest.length > 0) {
    const trimmed = rest.trimStart();
    if (trimmed.length !== rest.length) {
      rest = trimmed;
      continue;
    }
    if (rest.startsWith("{")) {
      const newlineIndex = rest.indexOf("\n");
      if (newlineIndex < 0) {
        return { messages, rest };
      }
      const line = rest.slice(0, newlineIndex).trim();
      rest = rest.slice(newlineIndex + 1);
      messages.push(line);
      continue;
    }
    const headerIndex = rest.indexOf(CONTENT_LENGTH_HEADER);
    if (headerIndex < 0) {
      return { messages, rest };
    }
    const afterHeader = rest.slice(headerIndex + CONTENT_LENGTH_HEADER.length);
    const headerEnd = afterHeader.indexOf("\r\n\r\n");
    if (headerEnd < 0) {
      return { messages, rest };
    }
    const lengthText = afterHeader.slice(0, afterHeader.indexOf("\r\n")).trim();
    const bodyLength = Number(lengthText);
    if (!Number.isInteger(bodyLength) || bodyLength < 0) {
      return { messages, rest: afterHeader.slice(headerEnd + 4) };
    }
    const bodyStart = headerIndex + CONTENT_LENGTH_HEADER.length + headerEnd + 4;
    if (rest.length < bodyStart + bodyLength) {
      return { messages, rest };
    }
    messages.push(rest.slice(bodyStart, bodyStart + bodyLength));
    rest = rest.slice(bodyStart + bodyLength);
  }
  return { messages, rest };
};

export const encodeStdioMessage = (body: string): string =>
  `${CONTENT_LENGTH_HEADER} ${Buffer.byteLength(body, "utf8")}\r\n\r\n${body}`;

export const forwardMcpJsonRpc = async (
  config: McpProxyConfig,
  body: string,
  fetchImpl: typeof fetch,
): Promise<string | null> => {
  const response = await fetchImpl(config.mcpUrl, {
    body,
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${config.mcpAuthToken}`,
      "CF-Access-Client-Id": config.accessClientId,
      "CF-Access-Client-Secret": config.accessClientSecret,
      "Content-Type": "application/json",
      "User-Agent": USER_AGENT,
    },
    method: "POST",
  });
  if (response.status === 202) {
    return null;
  }
  const text = await response.text();
  if (!response.ok) {
    return JSON.stringify({
      error: {
        code: -32000,
        message: `Remote MCP HTTP ${response.status}`,
      },
      id: null,
      jsonrpc: "2.0",
    });
  }
  if (text.trim().length === 0) {
    return null;
  }
  return text;
};

File: packages/pc-keiba-viewer-plugin/src/mcp_stdio_proxy.py
import json
from collections.abc import Mapping
from dataclasses import dataclass
from http.client import HTTPResponse
from typing import Protocol
from urllib.error import HTTPError, URLError
from urllib.parse import urlparse
from urllib.request import Request, urlopen

HTTPS_SCHEME: str = "https"
USER_AGENT: str = "pc-keiba-viewer-plugin/1.0"
CONTENT_LENGTH_HEADER: str = "Content-Length:"
HEADER_BODY_SEPARATOR: str = "\r\n\r\n"
HEADER_BODY_SEPARATOR_LENGTH: int = 4
ACCEPTED_STATUS: int = 202
JSON_RPC_HTTP_ERROR_CODE: int = -32000
HTTP_TIMEOUT_SECONDS: int = 60
UTF8: str = "utf-8"


@dataclass(frozen=True)
class McpProxyConfig:
    access_client_id: str | None
    access_client_secret: str | None
    mcp_auth_token: str
    mcp_url: str


@dataclass(frozen=True)
class ExtractedStdio:
    messages: tuple[str, ...]
    rest: str


@dataclass(frozen=True)
class McpHttpResponse:
    status: int
    text: str


class McpHttpPoster(Protocol):
    def __call__(self, *, url: str, body: str, headers: Mapping[str, str]) -> McpHttpResponse: ...


def load_mcp_proxy_config(env: Mapping[str, str]) -> McpProxyConfig | str:
    mcp_url = _trimmed_env(env, "PC_KEIBA_VIEWER_MCP_URL")
    mcp_auth_token = _trimmed_env(env, "MCP_AUTH_TOKEN")
    access_client_id = _trimmed_env(env, "PC_KEIBA_ACCESS_CLIENT_ID")
    access_client_secret = _trimmed_env(env, "PC_KEIBA_ACCESS_CLIENT_SECRET")
    if mcp_url is None:
        return "PC_KEIBA_VIEWER_MCP_URL is required (absolute https URL ending with /mcp)"
    if mcp_auth_token is None:
        return "MCP_AUTH_TOKEN is required"
    url_error = _validate_mcp_url(mcp_url)
    if url_error is not None:
        return url_error
    if access_client_id is None and access_client_secret is None:
        return McpProxyConfig(
            access_client_id=None,
            access_client_secret=None,
            mcp_auth_token=mcp_auth_token,
            mcp_url=mcp_url,
        )
    if access_client_id is None or access_client_secret is None:
        return "PC_KEIBA_ACCESS_CLIENT_ID and PC_KEIBA_ACCESS_CLIENT_SECRET must both be set or both omitted"
    return McpProxyConfig(
        access_client_id=access_client_id,
        access_client_secret=access_client_secret,
        mcp_auth_token=mcp_auth_token,
        mcp_url=mcp_url,
    )


def extract_stdio_messages(buffer: str) -> ExtractedStdio:
    stripped = buffer.lstrip()
    if stripped != buffer:
        return extract_stdio_messages(stripped)
    if buffer == "":
        return ExtractedStdio(messages=(), rest="")
    if buffer.startswith("{"):
        return _extract_ndjson_message(buffer)
    return _extract_content_length_message(buffer)


def encode_stdio_message(body: str) -> str:
    size = len(body.encode(UTF8))
    return f"{CONTENT_LENGTH_HEADER} {size}{HEADER_BODY_SEPARATOR}{body}"


def build_forward_headers(config: McpProxyConfig) -> dict[str, str]:
    headers: dict[str, str] = {
        "Accept": "application/json",
        "Authorization": f"Bearer {config.mcp_auth_token}",
        "Content-Type": "application/json",
        "User-Agent": USER_AGENT,
    }
    if config.access_client_id is None or config.access_client_secret is None:
        return headers
    headers["CF-Access-Client-Id"] = config.access_client_id
    headers["CF-Access-Client-Secret"] = config.access_client_secret
    return headers


def forward_mcp_jsonrpc(
    *,
    config: McpProxyConfig,
    body: str,
    poster: McpHttpPoster,
) -> str | None:
    try:
        response = poster(
            url=config.mcp_url,
            body=body,
            headers=build_forward_headers(config),
        )
    except URLError as error:
        return _json_rpc_http_error(f"Remote MCP transport error: {error.reason}")
    if response.status == ACCEPTED_STATUS:
        return None
    if response.status < 200 or response.status >= 300:
        return _json_rpc_http_error(f"Remote MCP HTTP {response.status}")
    if response.text.strip() == "":
        return None
    return response.text


def post_mcp_http(*, url: str, body: str, headers: Mapping[str, str]) -> McpHttpResponse:
    request = Request(
        url,
        data=body.encode(UTF8),
        headers=dict(headers),
        method="POST",
    )
    try:
        with urlopen(request, timeout=HTTP_TIMEOUT_SECONDS) as response:
            return _read_http_response(response)
    except HTTPError as error:
        return McpHttpResponse(status=error.code, text=error.read().decode(UTF8))


def _trimmed_env(env: Mapping[str, str], key: str) -> str | None:
    if key not in env:
        return None
    trimmed = env[key].strip()
    if trimmed == "":
        return None
    return trimmed


def _validate_mcp_url(mcp_url: str) -> str | None:
    parsed = urlparse(mcp_url)
    if parsed.scheme == "" or parsed.netloc == "":
        return "PC_KEIBA_VIEWER_MCP_URL is not a valid URL"
    if parsed.scheme != HTTPS_SCHEME:
        return "PC_KEIBA_VIEWER_MCP_URL must be https"
    if parsed.username is not None or parsed.password is not None:
        return "PC_KEIBA_VIEWER_MCP_URL must not include userinfo"
    if parsed.fragment != "":
        return "PC_KEIBA_VIEWER_MCP_URL must not include a fragment"
    if not parsed.path.endswith("/mcp"):
        return "PC_KEIBA_VIEWER_MCP_URL path must end with /mcp"
    return None


def _extract_ndjson_message(buffer: str) -> ExtractedStdio:
    newline_index = buffer.find("\n")
    if newline_index < 0:
        return ExtractedStdio(messages=(), rest=buffer)
    line = buffer[:newline_index].strip()
    following = extract_stdio_messages(buffer[newline_index + 1 :])
    return ExtractedStdio(messages=(line, *following.messages), rest=following.rest)


def _extract_content_length_message(buffer: str) -> ExtractedStdio:
    header_index = buffer.find(CONTENT_LENGTH_HEADER)
    if header_index < 0:
        return ExtractedStdio(messages=(), rest=buffer)
    after_header = buffer[header_index + len(CONTENT_LENGTH_HEADER) :]
    header_end = after_header.find(HEADER_BODY_SEPARATOR)
    if header_end < 0:
        return ExtractedStdio(messages=(), rest=buffer)
    line_end = after_header.find("\r\n")
    length_text = after_header[:line_end].strip()
    if not length_text.isdigit():
        return extract_stdio_messages(after_header[header_end + HEADER_BODY_SEPARATOR_LENGTH :])
    body_length = int(length_text)
    body_start = (
        header_index + len(CONTENT_LENGTH_HEADER) + header_end + HEADER_BODY_SEPARATOR_LENGTH
    )
    if len(buffer) < body_start + body_length:
        return ExtractedStdio(messages=(), rest=buffer)
    message = buffer[body_start : body_start + body_length]
    following = extract_stdio_messages(buffer[body_start + body_length :])
    return ExtractedStdio(messages=(message, *following.messages), rest=following.rest)


def _json_rpc_http_error(message: str) -> str:
    return json.dumps(
        {
            "error": {"code": JSON_RPC_HTTP_ERROR_CODE, "message": message},
            "id": None,
            "jsonrpc": "2.0",
        },
        separators=(",", ":"),
    )


def _read_http_response(response: HTTPResponse) -> McpHttpResponse:
    return McpHttpResponse(status=response.status, text=response.read().decode(UTF8))
